### PR TITLE
Fix/additional data from page source

### LINKED ISSUE #6189: Follow by tags doesn't work anymore
## Current Behavior
The follow button above a post has disappeared since 21/05. Which the bot uses to follow users.

## Possible Solution (optional)
Solution would be to use the follow button on a user his profile instead.

### LINKED ISSUE #6191: Like_by_tags gives error  media = post_page[0]["shortcode_media"] KeyError: 0
Hello All,

Facing issue while running like_by_tags function

Traceback (most recent call last):
  File "instagram.py", line 29, in <module>
    session.like_by_tags(['songs'],amount=2,interact=True,skip_top_posts=False)
  File "C:\Users\test\AppData\Local\Programs\Python\Python38\lib\site-packages\instapy\instapy.py", line 1939, in like_by_tags
    inappropriate, user_name, is_video, reason, scope = check_link(
  File "C:\Users\test\AppData\Local\Programs\Python\Python38\lib\site-packages\instapy\like_util.py", line 634, in check_link
    media = post_page[0]["shortcode_media"]
KeyError: 0

The code below in line 570 of like_util.py is not reading the post. The issue has reoccured which i think was fixed in 2019.
Value of variable post_page is {}

# Check if the Post is Valid/Exists
    try:
        post_page = browser.execute_script(
            "return window.__additionalData[Object.keys(window.__additionalData)[0]].data"
        )

### LINKED ISSUE #6194: like_util: post_page is getting empty response data
```
Traceback (most recent call last):
  File "quickstart.py", line 67, in <module>
    session.like_by_tags(['jokes','funny'], amount=100, interact=True)
  File "/home/nitesh/venv/lib/python3.8/site-packages/instapy/instapy.py", line 1957, in like_by_tags
    inappropriate, user_name, is_video, reason, scope = check_link(
  File "/home/nitesh/venv/lib/python3.8/site-packages/instapy/like_util.py", line 636, in check_link
    media = post_page[0]["shortcode_media"]
KeyError: 0
```

### LINKED ISSUE #6207: help
File "C:\Users\12146\AppData\Local\Programs\Python\Python39\lib\site-packages\instapy\instapy.py", line 1957, in like_by_tags
    inappropriate, user_name, is_video, reason, scope = check_link(
  File "C:\Users\12146\AppData\Local\Programs\Python\Python39\lib\site-packages\instapy\like_util.py", line 633, in check_link
    media = post_page[0]["shortcode_media"]
KeyError: 0

### LINKED ISSUE #6208: File "C:\Users\12146\Desktop\instagram bot\quickstart.py", line 32, in <module>     session.like_by_tags([ 'bongs' , 'waterpipe'], amount=100)   File "C:\Users\12146\AppData\Local\Programs\Python\Python39\lib\site-packages\instapy\instapy.py", line 1957, in like_by_tags     inappropriate, user_name, is_video, reason, scope = check_link(   File "C:\Users\12146\AppData\Local\Programs\Python\Python39\lib\site-packages\instapy\like_util.py", line 633, in check_link     media = post_page[0]["shortcode_media"]
File "C:\Users\12146\Desktop\instagram bot\quickstart.py", line 32, in <module>
    session.like_by_tags([ 'bongs' , 'waterpipe'], amount=100)
  File "C:\Users\12146\AppData\Local\Programs\Python\Python39\lib\site-packages\instapy\instapy.py", line 1957, in like_by_tags
    inappropriate, user_name, is_video, reason, scope = check_link(
  File "C:\Users\12146\AppData\Local\Programs\Python\Python39\lib\site-packages\instapy\like_util.py", line 633, in check_link
    media = post_page[0]["shortcode_media"]

### PR DESCRIPTION
**Pull request by schealex**
## Description

Add new function to utils to get additional data which was previously fetched from window object via js.

Fixes #6191
Fixes #6189
Fixes #6194
Fixes #6208
Fixes #6207

## How Has This Been Tested?

Was tested with my test instagram account and i can verify that at least liking by tag is working again now

## ToDo:

- [x] Check all places where the additional data is required and replace old js call with new function from utils

## Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have checked my code and corrected any misspellings
- [x] I have performed a self-review of my own code
- [x] My code follows the style guidelines of this project, `black -t py34`
- [x] My changes generate no new warnings

### CODE REVIEW

**Comment by schealex** [comment]
I've updated the PR so commenting should also work again, please test if you can guys

**Review by alokkumarsbg — COMMENTED** [review]
Changes look good. Working for me.

**Comment by nitesh585** [comment]
@schealex It's working for me 🔥

**Comment by techmaharaj** [comment]
Working fine for me too!

**Comment by schealex** [comment]
added comments, additional fixes and can confirm that everything is working again for me with those changes

**Comment by CharlesCCC** [comment]
if it working, should we get this PR merged ? @alokkumarsbg

**Comment by jlliao** [comment]
The file changes wouldn't work for interact_user_likers function. It completely messes up the scraping function for likers.

**Comment by caskednow** [comment]
@schealex works like a charm for me as well. Interact_by_url runs smooth now, thank you!

**Comment by schealex** [comment]
> The file changes wouldn't work for interact_user_likers function. It completely messes up the scraping function for likers.

can you give a sample configuration so i can try that out as i'm not using that functionality? I'm happy to test and solve the related problems if you supply me a config.

**Comment by alokkumarsbg** [comment]
> if it working, should we get this PR merged ? @alokkumarsbg

@CharlesCCC  Still answering the issue with others are facing. Let's all query get answered to merge a bug-free code

@jlliao interact_user_likers didn't gave me any error? Can you share your error or configuration you have used

**Review by alokkumarsbg — APPROVED** [review]
Code changes is working fine

**Review by alokkumarsbg — APPROVED** [review]
Reviewed

**Comment by schealex** [comment]
@breuerfelix / @analyticsdept is there someone with write access to review this?

and btw. is there still anyone with the access that checks here regularly? because there are a lot of open PR's and the issues are getting out of hand without administration i think

**Comment by rain82k** [comment]
Works 👍

**Comment by SanielDous** [comment]
Works for me too! Using `like_by_tags`

**Comment by analyticsdept** [comment]
> @breuerfelix / @analyticsdept is there someone with write access to review this?
> 
> and btw. is there still anyone with the access that checks here regularly? because there are a lot of open PR's and the issues are getting out of hand without administration i think

Hey @schealex - I'm working on macos these days and am really out of touch with the project so getting it up and running is a bit of a mission. Hopefully @breuerfelix can assist getting these PRs reviewed and accepted.

**Comment by yaelsprikut** [comment]
Using MacOS, can confirm that it's working for me too! 🎉

**Comment by Arucard11** [comment]
you are the goat thank you

**Comment by yaelsprikut** [comment]
I'm still getting `Cannot detect post media type. Skip ` on certain posts...

**Comment by SanielDous** [comment]
@yaelsprikut Try to modify this here too: https://github.com/timgrossmann/InstaPy/pull/6143/files

**Comment by tzvc** [comment]
Any updates on this?

**Comment by schealex** [comment]
@theochampion update about what? This PR is complete?!  Just waiting for someone to merge it

**Comment by schealex** [comment]
@SanielDous @yaelsprikut you can also just use my develop branch from my fork which has all the PRs and a couple more fixes in it. No guarantees though

**Comment by Arucard11** [comment]
It works perfect

**Comment by tzvc** [comment]
> @theochampion update about what? This PR is complete?!  Just waiting for someone to merge it 

Yes! This is what I was referring to in "update" ;)

**Comment by yaelsprikut** [comment]
> @yaelsprikut Try to modify this here too: https://github.com/timgrossmann/InstaPy/pull/6143/files

I actually tried this as well and it works for some posts, doesn't for others. It'll also sometimes just randomly crash on a post so I'm trying to figure out why that's happening.

**Comment by yaelsprikut** [comment]
> @SanielDous @yaelsprikut you can also just use my develop branch from my fork which has all the PRs and a couple more fixes in it. No guarantees though

It works but sporadically. Still better than not working at all though 😅

**Comment by yaelsprikut** [comment]
Ok so I've tested with [timgrossman's](https://github.com/timgrossmann/InstaPy/issues/6191#issuecomment-850329238) fix and it works great with most accounts, but I'm currently testing with an account that keeps failing with: 

`  File "/usr/local/lib/python3.8/site-packages/instapy/instapy.py", line 2555, in interact_by_users
    inappropriate, user_name, is_video, reason, scope = check_link(
  File "/usr/local/lib/python3.8/site-packages/instapy/like_util.py", line 574, in check_link
    post_page = get_additional_data(browser)
  File "/usr/local/lib/python3.8/site-packages/instapy/util.py", line 2577, in get_additional_data
    additional_data = json.loads(text[48:-2])
  File "/usr/local/Cellar/python@3.8/3.8.8_1/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/__init__.py", line 357, in loads
    return _default_decoder.decode(s)
  File "/usr/local/Cellar/python@3.8/3.8.8_1/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/local/Cellar/python@3.8/3.8.8_1/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)`

Has anyone else had this issue? I have a suspicion that this is only happening with private accounts even when you have access because you're following them.

File: instapy/comment_util.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
 # import InstaPy modules
 from .time_util import sleep
-from .util import update_activity
+from .util import update_activity, getMediaData
```
Comment (elulcao) [inline instapy/comment_util.py]: One import per line,

**Comment by elulcao** [comment]
CHANGELOG information is missing,
Could you change the PR to merge the change to dev branch,

**Comment by schealex** [comment]
> CHANGELOG information is missing,
> Could you change the PR to merge the change to dev branch,

@elulcao Fixed the import and added changelog info but why would you want to merge this to the dev branch? oO

**Review by elulcao — DISMISSED** [review]

**Comment by elulcao** [comment]
@schealex Thanks for creating this PR,
could you please check the Travis warning,
:)

**Comment by schealex** [comment]
> @schealex Thanks for creating this PR,
> could you please check the Travis warning,
> :)

i don't know whats the going on here but someone merged syntax errors into the master. Which is causing the errors here. Gonna fix it here since i want this to get merged finally.

**Comment by elulcao** [comment]
> > @schealex Thanks for creating this PR,
> > could you please check the Travis warning,
> > :)
> 
> i don't know whats the going on here but someone merged syntax errors into the master. Which is causing the errors here. Gonna fix it here since i want this to get merged finally.

I tried to fix the merge diff from previous commit, but web edit is not user-friendly. I guess something messed up with the indentations. btw, now all fine 
thanks a lot

**Review by elulcao — APPROVED** [review]

**Comment by yd47** [comment]
Hello! I have seen that function "liked_photo to set_do_comment to be able to comment twice a photo that is already liked; deactivated by default" have been added in release [0.6.13] - 2020-12-30. Does it work properly? In my case, if photo is already liked, the program does nothing, just notifies "Already liked". My code is:


```
from instapy import InstaPy
from instapy.util import smart_run


# login credentials
insta_username = '123'
insta_password = '123'

# get an InstaPy session!
# set headless_browser=True to run InstaPy in the background
session = InstaPy(username=insta_username,
                  password=insta_password,
                  headless_browser=False)

with smart_run(session):
    """ Activity flow """
    #define interaction settings
    #session.set_do_like(enabled=True, percentage=94)
    session.set_do_comment(enabled=True, percentage=100)
    session.set_comments(["test", "123"])

    
    #start the feature
    session.interact_by_URL(urls=["123"], randomize=False, interact=True)
```

**Comment by rajtulluri** [comment]
> 
> 
> I've updated the PR so commenting should also work again, please test if you can guys

Hi @schealex,
I am a beginner, just learning about this package. Could you tell me how I can install your version of the package?
Thanks a ton!

**Comment by prinaraujo** [comment]
hey guys, im new using instapy...

i used to get this issue, but i did everything in this chat.

but now im getting this one>

validation, details = validate_username(
TypeError: validate_username() missing 2 required positional arguments: 'logger' and 'logfolder'

could you guys help me

**Comment by lucasreq** [comment]
Hello ! 

I have a problem with graphql, there is an issue about it ? 
```
WARNING [2022-01-21 16:31:20] [testpython8]  Unable to save account progress, skipping data update 
        b"'graphql'"

KeyError: 'graphql'
```


Thanks !

**Pull request merged**